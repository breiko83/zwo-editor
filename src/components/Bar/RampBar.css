.trapeze {
  display: flex;
  align-items: flex-end;
}

.trapeze-component {  
  border-top: 1px dotted gray;
  z-index: 5;
}

.trapeze-background {
  position: absolute;
  left: 0;
  bottom: 0;
  width: 0;
  height: 0;
  background-color: blue;
}

.trapeze-colors {
  position: absolute;
  bottom: 0px;
  left: 0;
  width: 100%;
  display: flex;
}

.color {
  
}

.color:first-child{
  border-bottom-left-radius: 5px;
}

.color:last-child{
  border-bottom-right-radius: 5px;
}

.trapeze-svg-container {
  position: absolute;
  bottom: 0;
  left: 0;
}
.trapeze-svg {
  fill:transparent;
}
.trapeze-svg-off {
  fill:white;
  display: hidden;
}