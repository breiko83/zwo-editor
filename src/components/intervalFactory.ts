import { v4 as uuidv4 } from 'uuid'
import { PaceType } from './Editor/PaceSelector';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "./Interval";

const defaultDuration = 300;
const defaultPower = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

export default {
  steady(interval: Partial<SteadyInterval>): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      duration: defaultDuration,
      power: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      duration: defaultDuration,
      startPower: defaultPower / 2,
      endPower: defaultPower,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      duration: defaultDuration,
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      cadence: defaultCadence,
      restingCadence: defaultCadence,
      repeat: 3,
      onDuration: 30,
      offDuration: 120,
      onPower: defaultPower,
      offPower: defaultPower / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
