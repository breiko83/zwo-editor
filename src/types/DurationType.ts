export type DurationType = "time" | "distance";