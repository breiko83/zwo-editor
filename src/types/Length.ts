export class Duration {
  constructor(public readonly seconds: number) {}
}

export class Distance {
  constructor(public readonly meters: number) {}
}

export type Length = Distance | Duration;
