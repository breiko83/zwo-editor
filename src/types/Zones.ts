export enum ZoneColor {
  GRAY = '#807F80',
  BLUE = '#0E90D4',
  GREEN = '#00C46A',
  YELLOW = '#FFCB00',
  ORANGE = '#FF6430',
  RED = '#E90000',
  WHITE = '#FFFFFF',
}

export const Zones = {
  Z1: {
    min: 0.1,
    max: 0.605
  },
  Z2: {
    min: 0.605,
    max: 0.763
  },
  Z3: {
    min: 0.763,
    max: 0.901
  },
  Z4: {
    min: 0.901,
    max: 1.053
  },
  Z5: {
    min: 1.053,
    max: 1.194
  },
  Z6: {
    min: 1.194,
    max: 2.0
  }
}

export const ZonesArray = [
  [Zones.Z1.min,Zones.Z1.max],
  [Zones.Z2.min,Zones.Z2.max],
  [Zones.Z3.min,Zones.Z3.max],
  [Zones.Z4.min,Zones.Z4.max],
  [Zones.Z5.min,Zones.Z5.max],
  [Zones.Z6.min,Zones.Z6.max],
]

export const zoneColor = (intensity: number): ZoneColor => {
  if (intensity < Zones.Z1.max) {
    return ZoneColor.GRAY;
  } else if (intensity < Zones.Z2.max) {
    return ZoneColor.BLUE;
  } else if (intensity < Zones.Z3.max) {
    return ZoneColor.GREEN;
  } else if (intensity < Zones.Z4.max) {
    return ZoneColor.YELLOW;
  } else if (intensity < Zones.Z5.max) {
    return ZoneColor.ORANGE;
  } else {
    return ZoneColor.RED;
  }
};
