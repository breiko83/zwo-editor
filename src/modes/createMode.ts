import { LengthType } from "../types/LengthType";
import { RunningTimes } from "../types/RunningTimes";
import { SportType } from "../types/SportType";
import BikeMode from "./BikeMode";
import RunMode from "./RunMode";
import { WorkoutMode } from "./WorkoutMode";

interface CreateModeParams {
  sportType: SportType;
  ftp: number;
  weight: number;
  runningTimes: RunningTimes;
  lengthType: LengthType;
}

// Creates appropriate type of mode for current sport type
export default function createMode({sportType, ftp, weight, runningTimes, lengthType}: CreateModeParams): WorkoutMode {
  switch (sportType) {
    case "bike": return new BikeMode(ftp, weight);
    case "run": return new RunMode(runningTimes, lengthType);
  }
}
