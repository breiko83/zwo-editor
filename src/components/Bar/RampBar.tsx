import React, { useState } from 'react'
import './RampBar.css'
import { zoneColor, ZoneColor, Zones, ZonesArray } from '../../types/Zones'
import { Resizable } from 're-resizable'
import Label from '../Label/Label'
import { RampInterval } from '../../types/Interval'
import { intensityMultiplier } from './multipliers'
import { WorkoutMode } from '../../modes/WorkoutMode'

interface IDictionary {
  [index: string]: number;
}

interface RampBarProps {
  interval: RampInterval;
  mode: WorkoutMode;
  onChange: (interval: RampInterval) => void;
  onClick: (id: string) => void;
  selected: boolean;
}

const RampBar = ({interval, mode, ...props}: RampBarProps) => {
  const [showLabel, setShowLabel] = useState(false)

  const handleCadenceChange = (cadence: number) => {
    props.onChange({ ...interval, cadence: cadence })
  }

  const [width, setWidth] = useState(mode.lengthToWidth(interval.length))

  const [startHeight, setStartHeight] = useState(mode.intensityToHeight(interval.startIntensity))
  const [endHeight, setEndHeight] = useState(mode.intensityToHeight(interval.endIntensity))

  const handleStartResizeStop = (dHeight: number) => {    
    setStartHeight(startHeight + dHeight)
  }
  const handleEndResizeStop = (dWidth: number, dHeight: number) => {
    setWidth(width + dWidth)
    setEndHeight(endHeight + dHeight)
  }

  const handleStartResize = (dHeight: number) => {
    props.onChange({
      ...interval,
      startIntensity: mode.heightToIntensity(startHeight + dHeight),
    })
  }
  const handleEndResize = (dWidth: number, dHeight: number) => {
    props.onChange({
      ...interval,
      length: mode.widthToLength(width + dWidth),
      startIntensity: mode.heightToIntensity(startHeight),
      endIntensity: mode.heightToIntensity(endHeight + dHeight),
    })
  }

  return (
    <div className='segment'
      onMouseEnter={() => setShowLabel(true)}
      onMouseLeave={() => setShowLabel(false)}
      style={props.selected ? {zIndex:1}: {}}
      onClick={() => props.onClick(interval.id)}
    >
      {(props.selected || showLabel) &&
        <Label
          interval={interval}
          mode={mode}
          onCadenceChange={(cadence: number)=> handleCadenceChange(cadence)}
        />
      }
      <div className='trapeze' onClick={() => props.onClick(interval.id)}>
        <Resizable
          className='trapeze-component'
          size={{
            width: width / 2,
            height: startHeight,
          }}
          minWidth={3}
          minHeight={intensityMultiplier * Zones.Z1.min}
          maxHeight={intensityMultiplier * Zones.Z6.max}
          enable={{ top: true, right: true }}
          grid={[1, 1]}
          onResizeStop={(e, direction, ref, d) => handleStartResizeStop(d.height)}
          onResize={(e, direction, ref, d) => handleStartResize(d.height)}          
        >
        </Resizable>
        <Resizable
          className='trapeze-component'
          size={{
            width: width / 2,
            height: endHeight,
          }}
          minWidth={3}
          minHeight={intensityMultiplier * Zones.Z1.min}
          maxHeight={intensityMultiplier * Zones.Z6.max}
          enable={{ top: true, right: true }}
          grid={[1, 1]}
          onResizeStop={(e, direction, ref, d) => handleEndResizeStop(d.width, d.height)}
          onResize={(e, direction, ref, d) => handleEndResize(d.width, d.height)}
        >
        </Resizable>
      </div>
      <Rainbow interval={interval} startHeight={startHeight} endHeight={endHeight} />
      <SvgPolygons width={width} startHeight={startHeight} endHeight={endHeight} />
    </div>
  );
}

const Rainbow: React.FC<{interval: RampInterval, startHeight: number, endHeight: number}> = ({interval, startHeight, endHeight}) => {
  const trapezeHeight = endHeight > startHeight ? endHeight : startHeight;
  const bars = endHeight > startHeight ? calculateColors(interval.startIntensity, interval.endIntensity) : calculateColors(interval.endIntensity, interval.startIntensity);
  const flexDirection = endHeight > startHeight ? 'row' : 'row-reverse';

  return (
    <div className='trapeze-colors' style={{ height: trapezeHeight, flexDirection: flexDirection, backgroundColor: zoneColor(interval.startIntensity) }}>
      <div className='color' style={{ backgroundColor: ZoneColor.GRAY, width: `${(bars['Z1'] * 100 / Math.abs(interval.endIntensity - interval.startIntensity))}%` }}></div>
      <div className='color' style={{ backgroundColor: ZoneColor.BLUE, width: `${(bars['Z2'] * 100 / Math.abs(interval.endIntensity - interval.startIntensity))}%` }}></div>
      <div className='color' style={{ backgroundColor: ZoneColor.GREEN, width: `${(bars['Z3'] * 100 / Math.abs(interval.endIntensity - interval.startIntensity))}%` }}></div>
      <div className='color' style={{ backgroundColor: ZoneColor.YELLOW, width: `${(bars['Z4'] * 100 / Math.abs(interval.endIntensity - interval.startIntensity))}%` }}></div>
      <div className='color' style={{ backgroundColor: ZoneColor.ORANGE, width: `${(bars['Z5'] * 100 / Math.abs(interval.endIntensity - interval.startIntensity))}%` }}></div>
      <div className='color' style={{ backgroundColor: ZoneColor.RED, width: `${(bars['Z6'] * 100 / Math.abs(interval.endIntensity - interval.startIntensity))}%` }}></div>
    </div>
  );
};

function calculateColors(start: number, end: number): IDictionary {
  const bars = {} as IDictionary;

  ZonesArray.forEach((zone, index) => {
    if (start >= zone[0] && start < zone[1]) {
      bars['Z' + (index + 1)] = zone[1] - start
    }
    else if (end >= zone[0] && end < zone[1]) {
      bars['Z' + (index + 1)] = end - zone[0]
    }
    else if (end >= zone[1] && start < zone[0]) {
      bars['Z' + (index + 1)] = zone[1] - zone[0]
    } else {
      bars['Z' + (index + 1)] = 0
    }
  })
  return bars
}

const SvgPolygons: React.FC<{width: number, startHeight: number, endHeight: number}> = ({width, startHeight, endHeight}) => {
  const trapezeHeight = endHeight > startHeight ? endHeight : startHeight;
  const trapezeTop = endHeight > startHeight ? (endHeight - startHeight) : (startHeight - endHeight);

  const vertexB = endHeight > startHeight ? 0 : width;
  const vertexA = endHeight > startHeight ? trapezeTop : 0;
  const vertexD = endHeight > startHeight ? 0 : trapezeTop;

  return (
    <svg height={`${trapezeHeight}`} width={`${width}`} className='trapeze-svg-container'>
      <polygon points={`0,${vertexA} 0,${trapezeHeight} ${width},${trapezeHeight} ${width},${vertexD}`} className='trapeze-svg' />
      <polygon points={`0,0 ${vertexB},${trapezeTop} ${width},0`} className='trapeze-svg-off' />
    </svg>
  );
};

export default RampBar
