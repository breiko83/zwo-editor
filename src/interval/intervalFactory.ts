import { v4 as uuidv4 } from 'uuid'
import { PaceType } from '../types/PaceType';
import { FreeInterval, Interval, RampInterval, RepetitionInterval, SteadyInterval } from "../types/Interval";
import { Distance, Duration, Length } from '../types/Length';
import { WorkoutMode } from '../modes/WorkoutMode';

const defaultDuration = new Duration(300);
const defaultDistance = new Distance(1000);
const defaultIntensity = 1.0;
const defaultPace = PaceType.oneMile;
const defaultCadence = 0;

const defaultLength = (mode: WorkoutMode): Length =>
  mode.lengthType === "time" ? defaultDuration : defaultDistance;

export default {
  steady(interval: Partial<SteadyInterval>, mode: WorkoutMode): SteadyInterval {
    return {
      type: 'steady',
      id: uuidv4(),
      length: defaultLength(mode),
      intensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  ramp(interval: Partial<RampInterval>, mode: WorkoutMode): RampInterval {
    return {
      type: 'ramp',
      id: uuidv4(),
      length: defaultLength(mode),
      startIntensity: defaultIntensity / 2,
      endIntensity: defaultIntensity,
      cadence: defaultCadence,
      pace: defaultPace,
      ...interval,
    };
  },

  free(interval: Partial<FreeInterval>, mode: WorkoutMode): FreeInterval {
    return {
      type: 'free',
      id: uuidv4(),
      length: defaultLength(mode),
      cadence: defaultCadence,
      ...interval
    };
  },

  repetition(interval: Partial<RepetitionInterval>, mode: WorkoutMode): RepetitionInterval {
    return {
      type: 'repetition',
      id: uuidv4(),
      onCadence: defaultCadence,
      offCadence: defaultCadence,
      repeat: 3,
      onLength: mode.lengthType === "time" ? new Duration(30) : new Distance(200),
      offLength: mode.lengthType === "time" ? new Duration(120) : new Distance(200),
      onIntensity: defaultIntensity,
      offIntensity: defaultIntensity / 2,
      pace: defaultPace,
      ...interval,
    };
  },

  // Creates a copy of the interval, with a new ID
  clone(interval: Interval): Interval {
    return { ...interval, id: uuidv4() };
  },
};
