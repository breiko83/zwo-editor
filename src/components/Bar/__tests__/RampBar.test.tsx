import React from 'react';
import RampBar from '../RampBar';
import { Zones } from '../../../types/Zones'
import renderer from 'react-test-renderer';
import '@testing-library/jest-dom/extend-expect'
import intervalFactory from '../../../interval/intervalFactory';
import createMode from '../../../modes/createMode';
import { Duration } from '../../../types/Length';

const MockReact = React;

jest.mock('../../Label/Label', () => (props: any) =>
  MockReact.createElement("Label", props));

jest.mock('re-resizable', () => ({
  Resizable: (props: any) => MockReact.createElement("Resizable", props),
}));

jest.mock('uuid', () => ({
  v4: () => `mock-id`,
}));

describe('<SteadyBar>', () => {
  it('renders', () => {
    const mode = createMode({sportType: "bike", ftp: 250, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.ramp({
      length: new Duration(50),
      startIntensity: Zones.Z2.min,
      endIntensity: Zones.Z4.min,
    }, mode);

    const component = renderer.create(
      <RampBar
        interval={interval}
        mode={mode}
        selected={false}
        onChange={() => { }}
        onClick={() => { }}
      />
    )

    expect(component).toMatchSnapshot();
  });

  it('renders with Label when selected', () => {
    const mode = createMode({sportType: "bike", ftp: 250, weight: 75, runningTimes: [], lengthType: "time"});
    const interval = intervalFactory.ramp({
      length: new Duration(50),
      startIntensity: Zones.Z2.min,
      endIntensity: Zones.Z4.min,
    }, mode);

    const component = renderer.create(
      <RampBar
        interval={interval}
        mode={mode}
        selected={true}
        onChange={() => { }}
        onClick={() => { }}
      />
    )

    expect(component).toMatchSnapshot();
  });
});
