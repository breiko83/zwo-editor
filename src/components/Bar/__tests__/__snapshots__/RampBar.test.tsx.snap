// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SteadyBar> renders 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={Object {}}
>
  <div
    className="trapeze"
    onClick={[Function]}
  >
    <Resizable
      className="trapeze-component"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 151.25,
          "width": 8.333333333333334,
        }
      }
    />
    <Resizable
      className="trapeze-component"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 225.25,
          "width": 8.333333333333334,
        }
      }
    />
  </div>
  <div
    className="trapeze-colors"
    style={
      Object {
        "backgroundColor": "#0E90D4",
        "flexDirection": "row",
        "height": 225.25,
      }
    }
  >
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#807F80",
          "width": "0%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#0E90D4",
          "width": "53.37837837837838%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#00C46A",
          "width": "46.62162162162162%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#FFCB00",
          "width": "0%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#FF6430",
          "width": "0%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#E90000",
          "width": "0%",
        }
      }
    />
  </div>
  <svg
    className="trapeze-svg-container"
    height="225.25"
    width="16.666666666666668"
  >
    <polygon
      className="trapeze-svg"
      points="0,74 0,225.25 16.666666666666668,225.25 16.666666666666668,0"
    />
    <polygon
      className="trapeze-svg-off"
      points="0,0 0,74 16.666666666666668,0"
    />
  </svg>
</div>
`;

exports[`<SteadyBar> renders with Label when selected 1`] = `
<div
  className="segment"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  style={
    Object {
      "zIndex": 1,
    }
  }
>
  <Label
    interval={
      Object {
        "cadence": 0,
        "endIntensity": 0.901,
        "id": "mock-id",
        "length": Duration {
          "seconds": 50,
        },
        "pace": 0,
        "startIntensity": 0.605,
        "type": "ramp",
      }
    }
    mode={
      BikeMode {
        "ftp": 250,
        "lengthType": "time",
        "sportType": "bike",
        "weight": 75,
      }
    }
    onCadenceChange={[Function]}
  />
  <div
    className="trapeze"
    onClick={[Function]}
  >
    <Resizable
      className="trapeze-component"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 151.25,
          "width": 8.333333333333334,
        }
      }
    />
    <Resizable
      className="trapeze-component"
      enable={
        Object {
          "right": true,
          "top": true,
        }
      }
      grid={
        Array [
          1,
          1,
        ]
      }
      maxHeight={500}
      minHeight={25}
      minWidth={3}
      onResize={[Function]}
      onResizeStop={[Function]}
      size={
        Object {
          "height": 225.25,
          "width": 8.333333333333334,
        }
      }
    />
  </div>
  <div
    className="trapeze-colors"
    style={
      Object {
        "backgroundColor": "#0E90D4",
        "flexDirection": "row",
        "height": 225.25,
      }
    }
  >
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#807F80",
          "width": "0%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#0E90D4",
          "width": "53.37837837837838%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#00C46A",
          "width": "46.62162162162162%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#FFCB00",
          "width": "0%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#FF6430",
          "width": "0%",
        }
      }
    />
    <div
      className="color"
      style={
        Object {
          "backgroundColor": "#E90000",
          "width": "0%",
        }
      }
    />
  </div>
  <svg
    className="trapeze-svg-container"
    height="225.25"
    width="16.666666666666668"
  >
    <polygon
      className="trapeze-svg"
      points="0,74 0,225.25 16.666666666666668,225.25 16.666666666666668,0"
    />
    <polygon
      className="trapeze-svg-off"
      points="0,0 0,74 16.666666666666668,0"
    />
  </svg>
</div>
`;
